### ISSUE TITLE
Peel off the first iteration(s) of a loop in order to stabilize profiling information

### ISSUE THREAD

**Issue by Krovatkin** [pytorch/pytorch#38816]
## 🚀 Feature
The tensor values (phi values) in a loop sometimes have unstable `requires_grad` properties that change from not required to required.
This behaviour occurs because these values are recomputed by using parameters/weights which be do require gradients by default.
We would like to peel of the first iteraton(s) of a loop in order to stabilize the `required_grad` property and potentially other properties.
In the example below the initial values for `hy, cy` don't require gradients, however in order to recompute them, `self.weight_ih`, etc are used, so in the next iteration `hy, hc` will require gradients.

```python
    @jit.script_method
    def forward(self, input, state):
        # type: (Tensor, Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tuple[Tensor, Tensor]]
        hx, cx = state
        gates = (torch.mm(input, self.weight_ih.t()) + self.bias_ih +
                 torch.mm(hx, self.weight_hh.t()) + self.bias_hh)
        ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)

        ingate = torch.sigmoid(ingate)
        forgetgate = torch.sigmoid(forgetgate)
        cellgate = torch.tanh(cellgate)
        outgate = torch.sigmoid(outgate)

        cy = (forgetgate * cx) + (ingate * cellgate)
        hy = outgate * torch.tanh(cy)

        return hy, (hy, cy)
```

```
[DUMP guard_elimination.cpp:022]   %hy : Tensor, %cy : Tensor = prim::Loop(%12, %7, %73, %74) # test.py:26:4
[DUMP guard_elimination.cpp:022]     block0(%seq_idx.1 : int, %hy.5 : Tensor, %cy.5 : Tensor):
[DUMP guard_elimination.cpp:022]       %93 : Float(64:512, 512:1) = prim::Guard(%cy.5)
[DUMP guard_elimination.cpp:022]       %79 : Float(64:512, 512:1) = prim::Guard(%hy.5)
[DUMP guard_elimination.cpp:022]       %86 : Float(2048:1) = prim::Guard(%bhh.1)
[DUMP guard_elimination.cpp:022]       %84 : Float(2048:1) = prim::Guard(%bih.1)
[DUMP guard_elimination.cpp:022]       %78 : Float(2048:512, 512:1) = prim::Guard(%whh.1)
[DUMP guard_elimination.cpp:022]       %75 : Float(2048:512, 512:1) = prim::Guard(%wih.1)
[DUMP guard_elimination.cpp:022]       %20 : Tensor = aten::__getitem__(%inputs.1, %seq_idx.1) # test.py:27:32
[DUMP guard_elimination.cpp:022]       %76 : Float(64:512, 512:1) = prim::Guard(%20)
[DUMP guard_elimination.cpp:022]       %22 : Tensor = aten::t(%75) # test.py:7:28
[DUMP guard_elimination.cpp:022]       %77 : Float(512:1, 2048:512) = prim::Guard(%22)
[DUMP guard_elimination.cpp:022]       %25 : Tensor = aten::mm(%76, %77) # test.py:7:12
[DUMP guard_elimination.cpp:022]       %81 : Float(64:2048, 2048:1) = prim::Guard(%25)
[DUMP guard_elimination.cpp:022]       %27 : Tensor = aten::t(%78) # test.py:7:53
[DUMP guard_elimination.cpp:022]       %80 : Float(512:1, 2048:512) = prim::Guard(%27)
[DUMP guard_elimination.cpp:022]       %30 : Tensor = aten::mm(%79, %80) # test.py:7:40
[DUMP guard_elimination.cpp:022]       %82 : Float(64:2048, 2048:1) = prim::Guard(%30)
[DUMP guard_elimination.cpp:022]       %33 : Tensor = aten::add(%81, %82, %9) # test.py:7:12
[DUMP guard_elimination.cpp:022]       %83 : Float(64:2048, 2048:1) = prim::Guard(%33)
[DUMP guard_elimination.cpp:022]       %36 : Tensor = aten::add(%83, %84, %9) # test.py:7:12
[DUMP guard_elimination.cpp:022]       %85 : Float(64:2048, 2048:1) = prim::Guard(%36)
[DUMP guard_elimination.cpp:022]       %gates.1 : Tensor = aten::add(%85, %86, %9) # test.py:7:12
[DUMP guard_elimination.cpp:022]       %87 : Float(64:2048, 2048:1) = prim::Guard(%gates.1)
[DUMP guard_elimination.cpp:022]       %41 : Tensor, %42 : Tensor, %43 : Tensor, %44 : Tensor = prim::ConstantChunk[chunks=4, dim=1](%87)
[DUMP guard_elimination.cpp:022]       %91 : Float(64:2048, 512:1) = prim::Guard(%44)
[DUMP guard_elimination.cpp:022]       %90 : Float(64:2048, 512:1) = prim::Guard(%43)
[DUMP guard_elimination.cpp:022]       %89 : Float(64:2048, 512:1) = prim::Guard(%42)
[DUMP guard_elimination.cpp:022]       %88 : Float(64:2048, 512:1) = prim::Guard(%41)
[DUMP guard_elimination.cpp:022]       %ingate.3 : Tensor = aten::sigmoid(%88) # test.py:11:13
[DUMP guard_elimination.cpp:022]       %94 : Float(64:512, 512:1) = prim::Guard(%ingate.3)
[DUMP guard_elimination.cpp:022]       %forgetgate.3 : Tensor = aten::sigmoid(%89) # test.py:12:17
[DUMP guard_elimination.cpp:022]       %92 : Float(64:512, 512:1) = prim::Guard(%forgetgate.3)
[DUMP guard_elimination.cpp:022]       %cellgate.3 : Tensor = aten::tanh(%90) # test.py:13:15
[DUMP guard_elimination.cpp:022]       %95 : Float(64:512, 512:1) = prim::Guard(%cellgate.3)
[DUMP guard_elimination.cpp:022]       %outgate.3 : Tensor = aten::sigmoid(%91) # test.py:14:14
[DUMP guard_elimination.cpp:022]       %99 : Float(64:512, 512:1) = prim::Guard(%outgate.3)
[DUMP guard_elimination.cpp:022]       %55 : Tensor = aten::mul(%92, %93) # test.py:16:10
[DUMP guard_elimination.cpp:022]       %96 : Float(64:512, 512:1) = prim::Guard(%55)
[DUMP guard_elimination.cpp:022]       %58 : Tensor = aten::mul(%94, %95) # test.py:16:30
[DUMP guard_elimination.cpp:022]       %97 : Float(64:512, 512:1) = prim::Guard(%58)
[DUMP guard_elimination.cpp:022]       %cy.1 : Tensor = aten::add(%96, %97, %9) # test.py:16:10
[DUMP guard_elimination.cpp:022]       %102 : Float(64:512, 512:1) = prim::Guard(%cy.1)
[DUMP guard_elimination.cpp:022]       %98 : Float(64:512, 512:1) = prim::Guard(%cy.1)
[DUMP guard_elimination.cpp:022]       %63 : Tensor = aten::tanh(%98) # test.py:17:19
[DUMP guard_elimination.cpp:022]       %100 : Float(64:512, 512:1) = prim::Guard(%63)
[DUMP guard_elimination.cpp:022]       %hy.1 : Tensor = aten::mul(%99, %100) # test.py:17:9
[DUMP guard_elimination.cpp:022]       %101 : Float(64:512, 512:1) = prim::Guard(%hy.1)
[DUMP guard_elimination.cpp:022]       -> (%7, %101, %102)
```

## Motivation

The unstable properties interfere with guard elimination and eventually with fusion leading to suboptimal performance.

## Pitch

* Add a new pass that would peel off the first N (configurable) iteration of loops in a graph before it's instrumented with `prim::profile` nodes.

## Alternatives

Keep rewriting profiling information in every loop iteration if the number of profiling runs is equal to 1.

## Additional context
A full guard elimination log file.
[guard_elimination.log](https://github.com/pytorch/pytorch/files/4659271/guard_elimination.log)
